controllers/core/openstackversion_controller.go

```diff
@@ -0,0 +1,310 @@
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers. Return and don't requeue.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always patch the instance status when exiting this function so we can persist any changes.
+	defer func() {
+		/*
+		   // update the Ready condition based on the sub conditions
+		   if instance.Status.Conditions.AllSubConditionIsTrue() {
+		       instance.Status.Conditions.MarkTrue(
+		           condition.ReadyCondition, condition.ReadyMessage)
+		   } else {
+		       // something is not ready so reset the Ready condition
+		       instance.Status.Conditions.MarkUnknown(
+		           condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
+		       // and recalculate it based on the state of the rest of the conditions
+		       instance.Status.Conditions.Set(
+		           instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		   }*/
+
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+	instance.Status.AvailableVersion = envOpenStackVersion
+	instance.Status.AvailableServices = make([]string, 0)
+	for service, _ := range envVarsFromMain {
+		instance.Status.AvailableServices = append(instance.Status.AvailableServices, (strings.TrimPrefix(strings.TrimSuffix(service, "_IMAGE_URL_DEFAULT"), "RELATED_IMAGE_")))
+	}
+
+	if len(instance.Spec.TargetVersion) > 0 && instance.Spec.TargetVersion == instance.Status.TargetVersion {
+		return ctrl.Result{}, nil
+	}
+
+	// lookup the named OpenStackControlplane
+	controlPlane, err := r.lookupOpenStackControlPlane(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// services
+	glance := CheckGlanceImages(ctx, Log, &controlPlane.Spec.Glance.Template, instance.Spec.ServiceExcludes, envVarsFromMain)
+	if !glance {
+		instance.Status.ServicesNeedingUpdates = appendServiceIfMissing(instance.Status.ServicesNeedingUpdates, corev1beta1.OpenStackService{ServiceName: "glance"})
+	}
+
+	cinder := CheckCinderImages(ctx, Log, &controlPlane.Spec.Cinder.Template, instance.Spec.ServiceExcludes, instance.Spec.CinderVolumeExcludes, envVarsFromMain)
+	if !cinder {
+		instance.Status.ServicesNeedingUpdates = appendServiceIfMissing(instance.Status.ServicesNeedingUpdates, corev1beta1.OpenStackService{ServiceName: "cinder"})
+	}
+
+	if instance.Spec.TargetVersion != envOpenStackVersion {
+		return ctrl.Result{}, fmt.Errorf(fmt.Sprintf("Target version must equal: %s", envOpenStackVersion))
```

Comment:
I would move this to before the service specific checks or even before the r.lookupOpenStackControlPlane() as if the requested target is wrong then we don't need to do all that work. 
//later
Or move it to a validation webhook to provide direct feedback to the user that the specified value is wrong

Reply:
Agree, just haven't gotten to the webhooks work yet. I will add a comment to this effect